### PR TITLE
Create cli tools...and we support all SPLToken instructions now.

### PR DESCRIPTION
**Pull request by czl1378**
#1 #2

### CODE REVIEW

File: bin/utils.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
+var shell = require('shelljs');
```
Comment (defactojob) [inline bin/utils.js:1]: let's avoid adding a dependency? 

```js
const util = require('util');
const exec = util.promisify(require('child_process').exec);

async function lsExample() {
  const { stdout, stderr } = await exec('ls');
  console.log('stdout:', stdout);
  console.error('stderr:', stderr);
}
lsExample();

```

File: bin/utils.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+var programBuilder = path.join(bpfSdkDir, 'rust/build.sh');
+
+function exec(cmd, fn) {
+  shell.exec(cmd, function (code, stdout) {
+    if (code === 0) {
+      fn(stdout)
+    } else {
+      console.log('Some error occurs, exit.')
+    }
+  });
```
Comment (defactojob) [inline bin/utils.js:24]: use promise, not callback

File: bin/utils.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      // Find the first .so file in the directory
+      var files = fs.readdirSync(profilePath);
+      var soFile = '';
+      files.every(function(file) {
+        return !(/.*\.so$/.test(file) && (soFile = file));
+      });
+      return path.join(profilePath, soFile);
+    }
+  } catch(e) {
+    return '';
```
Comment (defactojob) [inline bin/utils.js:45]: does it make sense to return '' here? maybe just let error propagate?

File: bin/utils.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Build the solana program.
+function runBuild(program, toPath) {
+
+  var programDir = path.join(cwd, program);
+  var targetDir = path.join(programDir, 'target');
+  var profilePath = path.join(targetDir, 'bpfel-unknown-unknown', 'release');
+ 
+  exec(programBuilder + ' ' + program, function(res) {
```
Comment (defactojob) [inline bin/utils.js:56]: ban callback API.

File: bin/utils.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        var res = shell.cp('-f', soFilePath, outputFilePath);
+         
+        if (res.stderr) {
+          throw new Error();
+        } else {
+          console.log('Program built to:', outputFilePath);
+        }
+
+      } catch(err) {
+        console.log('Move *.so file error, maybe you can find the file in', profilePath);
```
Comment (defactojob) [inline bin/utils.js:84]: please make the try / catch scope as narrow as possible.

File: bin/utils.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+      }
+    } else {
+      console.log('Get built *.so file error, exit.');
```
Comment (defactojob) [inline bin/utils.js:87]: `else` very far away from the if...

in general, avoid if nestings. Check for error conditions, then immediately return:

```
if (!soFilePath) {
   console.log('Get built *.so file error, exit.');
   return;
}

// normal code flow follows...
```

File: bin/utils.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  var sdkVersionData = fs.readFileSync(sdkVersionFile, 'utf-8');
+  var pkgJSONData = fs.readFileSync(path.join(rootDir, 'package.json'), 'utf-8');
+
+  var pkgJSON = JSON.parse(pkgJSONData) || {};
+  
+  return pkgJSON.testnetDefaultChannel == 'v' + (/^(.*)\s/.exec(sdkVersionData) || [])[1];
+}
+
+exports.build = function(program, toPath) {
+  if (!fs.existsSync(bpfSdkDir)) {
```
Comment (defactojob) [inline bin/utils.js:103]: turn this into async/await

**Review by defactojob — CHANGES_REQUESTED** [review]

**Comment by defactojob** [comment]
let's convert the bin/utils.js to typescript. put it in `src/bin/utils.ts`

File: src/bin/utils.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    } else {
+      console.log('SDK version not match, download the new version.');
+      fs.rmdirSync(bpfSdkDir, { recursive: true });
+    }
+  }
+  try {
+    // We need install sdk first
+    await exec(sdkInstaller + ' ' + web3Dir);
+    await runBuild(program, toPath);
+  } catch(err){};
```
Comment (defactojob) [inline src/bin/utils.ts:107]: simpler:

```
await ensureLatestBuildSDK(bpfSdkDir)
await runBuild(program, toPath);
```

**Review by defactojob — COMMENTED** [review]
refactor 

```
export async function build(program: string, toPath: string): Promise<void> {
```

**Comment by czl1378** [comment]
It's okay now. Thanks for your review~

**Pull request merged**